File: medium/validate-binary-search-tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode], min=None, max=None) -> bool:
        if root == None:
            return True
        val = root.val
        if min != None and val >= min:
            return False
        if max != None and val <= max:
            return False
        return self.isValidBST(root.left, val, max) and self.isValidBST(root.right, min, val)

File: medium/validate-binary-search-tree/validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function (root, min, max) {
  if (!root) {
    return true;
  }
  const val = root.val;
  if (min !== undefined && val >= min) {
    return false;
  }
  if (max !== undefined && val <= max) {
    return false;
  }
  return isValidBST(root.left, val, max) && isValidBST(root.right, min, val);
};
